### PR TITLE
[TorchToLinalg] : Lower count_nonzero to Linalg-on-Tensors

### PR DESCRIPTION
**Pull request by um-urosmarkovic**
- Lowered count_nonzero op to Linalg-on-Tensors
- Decomposed aten.count_nonzero using aten.ne and aten.ne.Scalar and aten.sum/aten.sum.dim_IntList
- Added tests to torch-mlir/projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py

### CODE REVIEW

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1 , could you please approve the workflows when you get the chance? Thanks!

**Comment by um-urosmarkovic** [comment]
Hi @vivekkhandelwal1 , I’ve rebased the branch and the CI workflows need maintainer approval again. Could you please click “Approve and run” when you have a moment? Thanks!

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 1]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+    @export
+    @annotate_args([None, ([2, 3, 4], torch.int64, True)])
+    def forward(self, x):
+        return torch.ops.aten.count_nonzero(x, self.dim)
+
+
+@register_test_case(module_factory=lambda: CountNonzeroModuleI64())
+def CountNonzeroModuleI64_basic(module, tu: TestUtils):
+    module.forward(tu.randint(2, 3, 4, low=-2, high=2)).to(torch.int64)
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2517]: You don't need to do `to(torch.int64)` since it generates a tensor with `dtype=torch.int64`.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 2]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+    @export
+    @annotate_args([None, ([2, 3, 4], torch.float32, True)])
+    def forward(self, x):
+        return torch.ops.aten.count_nonzero(x, self.dim)
+
+
+@register_test_case(module_factory=lambda: CountNonzeroModuleF32())
+def CountNonzeroModuleF32_basic(module, tu: TestUtils):
+    module.forward(tu.rand(2, 3, 4).to(torch.float32))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2533]: Same here as above.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 3]
```diff
@@ -2496,3 +2496,54 @@ def TraceUnsignedIntModule_basic(module, tu: TestUtils):
...
+
+
+class CountNonzeroModuleBool(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.dim = -3
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py:2539]: It's not used in this test. Please remove it.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -6661,6 +6661,46 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    : public OpRewritePattern<AtenCountNonzeroOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenCountNonzeroOp op,
+                                PatternRewriter &rewriter) const override {
+    auto dim = op.getDim();
+    if (!isa<Torch::NoneType>(dim.getType()) &&
+        !isa<Torch::IntType>(dim.getType())) {
+      return rewriter.notifyMatchFailure(op,
+                                         "expected `dim` to be `None` or `int` "
+                                         "or constructed from list construct");
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6677]: Why do you expect dim to be "constructed from list construct"?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -6661,6 +6661,46 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    }
+    Location loc = op.getLoc();
+    auto self = op.getSelf();
+    auto inTy = cast<BaseTensorType>(self.getType());
+    auto boolTenTy =
+        inTy.getWithSizesAndDtype(inTy.getSizes(), rewriter.getI1Type());
+    auto C_0 =
+        rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6685]: Please follow the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly) for the variable names. Rename it to something like: `cstZeroInt` or `cstIntZero`, or `cstZero`.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -6661,6 +6661,46 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+        !isa<Torch::IntType>(dim.getType())) {
+      return rewriter.notifyMatchFailure(op,
+                                         "expected `dim` to be `None` or `int` "
+                                         "or constructed from list construct");
+    }
+    Location loc = op.getLoc();
+    auto self = op.getSelf();
+    auto inTy = cast<BaseTensorType>(self.getType());
+    auto boolTenTy =
+        inTy.getWithSizesAndDtype(inTy.getSizes(), rewriter.getI1Type());
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6683]: `boolTenTy` -> `inpBoolTy` or `neResType`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -6661,6 +6661,46 @@ class DecomposeAtenRot90Op : public OpRewritePattern<AtenRot90Op> {
...
+    } else {
+      auto C_False = rewriter.create<ConstantBoolOp>(loc, false);
+      auto dimIntList = rewriter.create<PrimListConstructOp>(
+          loc, ListType::get(IntType::get(op.getContext())),
+          SmallVector<Value>{dim});
+      rewriter.replaceOpWithNewOp<AtenSumDimIntListOp>(
+          op, op.getResult().getType(), Ne_0, dimIntList, C_False, none);
+    }
+    return success();
+  }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:6700]: Similarly, make the variable names relevant and meaningful here.

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 8]
```diff
@@ -6033,6 +6033,32 @@ LogicalResult AtenRot90Op::verify() {
...
+LogicalResult AtenCountNonzeroOp::verify() {
+
+  auto selfType = cast<BaseTensorType>(getSelf().getType());
+
+  if (!selfType.hasDtype() || !selfType.hasSizes())
+    return success();
+
+  int64_t dim;
+  if (!matchPattern(getDim(), m_TorchConstantInt(&dim)))
+    return success();
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/IR/TorchOps.cpp:6049]: This code should be placed after the below condition check for the dim type.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

**Comment by um-urosmarkovic** [comment]
Thanks for the review, @vivekkhandelwal1! I think I have addressed all requested changes—please let me know if there’s anything else to fix.

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Comment by um-urosmarkovic** [comment]
Thanks @vivekkhandelwal1 ! Could you, please, help me merge it? I don’t have the permissions. Much appreciated!
